Refactor subscriptions code so it can be used in Fusion

roccotripaldi opened the pull request:
Fixes a problem we've seen in developing the subscription block where we sometimes
forget to make the equivalent upstream changes.

#### Changes proposed in this Pull Request:

* Moves the class Jetpack_Subscriptions_Widget into /modules/subscriptions/views.php
* Moves the registration for the subscriptions block into /modules/subscriptions/views.php
* /modules/subscriptions/views.php will be shared with wordpress.com

#### Testing instructions:

- Get this branch running locally or on jurassic.ninja
- Ensure subscription widgets still function as usual
- add a widget to the blog
- add a subscription block to a post
- try subscribing from the front end

#### Proposed changelog entry for your changes:
No changelog needed

roccotripaldi committed:
Refactor subscriptions code so it can be used in Fusion

jeherve reviewed (changes requested):
I'm seeing some warnings when trying to use the forms on a test site:

```
Strict Standards: Non-static method Jetpack_Subscriptions_Widget::defaults() should not be called statically in wp-content/plugins/jetpack-dev/modules/subscriptions/views.php on line 640

Notice: Undefined index: subscribe_logged_in in wp-content/plugins/jetpack-dev/modules/subscriptions/views.php on line 227
```

roccotripaldi committed:
address feedback from @jeherve

jeherve reviewed (changes requested):
I added a few more comments.

I think it may be worth looking at the output from `phpcs` in your editor on the new changes; it helps and gives quite a few recommendations.

There is also a mix of spaces added all over the new view file, instead of tabs. Your editor's linter should be able to clean things up a bit.

Overall, it makes me wonder if it would not be easier to get rid of as much duplication as possible, and instead make a few changes to the WordPress.com widget to unify things a bit. We basically have 2 different interfaces in one file right now, that try to do the same thing, and used to do the same thing, but have not been kept in sync so far.

roccotripaldi commented on modules/subscriptions/views.php:
```diff
@@ -0,0 +1,670 @@
+			 * @since 2.1.2
+			 *
+			 * @param string $widget_name Widget title.
+			 */
```
will that make sense to someone browsing this file on the wpcom code base?

dereksmart replied:
@roccotripaldi maybe not, but these docblocks are for developer.jetpack.com - writing the comment as Jeremy suggested will ensure there aren't duplicate entries on that site, and it follows the convention of all other filter docblocks.

roccotripaldi replied:
ah, thanks for clarifying

roccotripaldi commented on modules/subscriptions/views.php:
```diff
@@ -0,0 +1,670 @@
+			}
+		}
+
+		if ( self::is_jetpack() ) {
```
Thanks for explaining this!

roccotripaldi commented on modules/subscriptions/views.php:
```diff
@@ -0,0 +1,670 @@
+
+		if ( self::is_jetpack() && empty( $instance['show_only_email_and_button'] ) ) {
+			echo $args['before_title'] . esc_attr( $instance['title'] ) . $args['after_title'] . "\n";
+		}
```
No fall back needed - in the case of Jetpack but with $instance['show_only_email_and_button'] set - we will do nothing.

I'm adding a docblock that should help clear this up.

roccotripaldi committed:
adding docblocks, and adjusting spacing / indents
fixing some logistics

thanks for the suggestions @jeherve

roccotripaldi commented on modules/subscriptions/views.php:
```diff
@@ -0,0 +1,670 @@
+		if ( self::is_wpcom() && ! $show_only_email_and_button ) {
+			if ( self::is_current_user_subscribed() ) {
+				if ( ! empty( $instance['title_following'] ) ) {
+					echo $before_title . '<label for="subscribe-field' . ( Jetpack_Subscriptions_Widget::$instance_count > 1 ? '-' . Jetpack_Subscriptions_Widget::$instance_count : '' ) . '">' . esc_attr( $instance['title_following'] ) . '</label>' . $after_title . "\n";
```
I don't know how to fix this or make it more clear.

dereksmart replied:
This is fine @jeherve

roccotripaldi commented on modules/subscriptions/views.php:
```diff
@@ -0,0 +1,670 @@
+				?>
+                <p><?php printf( __( 'You are following this blog (<a href="%s">manage</a>).' ), $edit_subs_url ) ?></p><?php
+			endif;
+		}
```
I'll leave it as is - it's nice to see the difference between the code bases
In wpcom we can easily know if someone is following the blog, so we show a helpful message.
In jetpack, it's not as easy to know, so we do nothing.

jeherve commented on modules/subscriptions/views.php:
```diff
@@ -0,0 +1,754 @@
+
+		if ( self::is_jetpack() ) {
+			$defaults['title']                 = esc_html__( 'Subscribe to Blog via Email', 'jetpack' );
+			$defaults['subscribe_text']        = esc_html__( 'Subscribe to Blog via Email', 'jetpack' );
```
Those default values do not appear to be correct; they are all the same

roccotripaldi replied:
great catch!

jeherve commented on modules/subscriptions/views.php:
```diff
@@ -0,0 +1,754 @@
+	function enqueue_style() {
+		wp_register_style(
+			'jetpack-subscriptions',
+			plugins_url( 'subscriptions/subscriptions.css', __FILE__ ),
```
This won't work in Jetpack; it tries to load `wp-content/plugins/jetpack/modules/subscriptions/subscriptions/subscriptions.css?ver=6.9-alpha`

jeherve reviewed (changes requested):
I spotted a couple of additional issues that it would be nice to fix as well.

roccotripaldi committed:
fix defaults to use their previous strings

roccotripaldi committed:
use correct url for style enqueueing

roccotripaldi commented:
thanks for the reviews @jeherve - i addressed your feedback in a few recent commits.

today we are going to try to merge the .com side, and if all goes well there, i'd say we are ready to merge this.

see D22567-code

roccotripaldi committed:
remove duplicated doc-blocks

jeherve reviewed (approved):
Everything seems to work right now. I think we have the potential to do a lot more to simplify all this in the future and get rid of most of the PHPCS errors that old codebase generates, but as you mentioned it's a good start to bring the 2 codebases together. 👍

dereksmart commented:
+1, this tests well for me and is a good step in the right direction. @roccotripaldi I left one comment in the phab diff regarding checking with VIP themes before merging to wpcom, but aside from that I feel this is good to go.
